File: algorithms/D/Distinct Subsequences/Distinct Subsequences.cpp
class Solution {
public:
    int solve(string &s, string &t, int i, int j, vector<vector<int>> &dp){
        if(j >= t.size()) return 1;
        if(i >= s.size() || t.size() - j > s.size() - i) return 0;
        if(dp[i][j] != -1) return dp[i][j];
        int res = 0;
        for(int k = i; k < s.size(); k++){
            if(s[k] == t[j]) res += solve(s, t, k+1, j+1, dp); 
        }
        return dp[i][j] = res;
    }
    
    int numDistinct(string s, string t) {
        if(s.size() == t.size()) return s == t;
        vector<vector<int>> dp(s.size(), vector<int> (t.size(), -1));
        return solve(s, t, 0 , 0, dp);
    }
};

File: scripts/algorithms/D/Distinct Subsequences/Distinct Subsequences.py
# Runtime: 1092 ms (Top 41.44%) | Memory: 70.6 MB (Top 63.39%)
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        if len(t) > len(s):
            return 0
        ls, lt = len(s), len(t)
        res = 0
        dp = [[0] * (ls + 1) for _ in range(lt + 1)]
        for j in range(ls + 1):
            dp[-1][j] = 1
        for i in range(lt - 1, -1, -1):
            for j in range(ls -1 , -1, -1):
                dp[i][j] = dp[i][j + 1]
                if s[j] == t[i]:

                    dp[i][j] += dp[i + 1][j + 1]
        return dp[0][0]